### PR TITLE
Impro-1121: Rename probability cubes - 1

### PR DESCRIPTION
**Pull request by markysparks**
Addresses #GitHubissuenum

Rename all IMPROVER cubes that contain probabilities like  `probability_of_rainfall_rate` to follow a specified convention that includes whether the probability is above or below a threshold within the cube name, so that the cube name better describes the contents of the cube. For example:

- `probability_of_rainfall_rate_above_threshold`
- `probability_of_visibility_in_air_below_threshold`

Testing:
 - [x] Ran tests and they passed OK
 - [x] Unit and CLI tests updated as required.

This initial PR only includes changes to the following areas:
```
- wxcodes (the greatest number of changes)
- threshold
- set_up_test_cubes
- cube_metadata
- weighted_blend
- nowcast lightning
```
A subsequent PR will be used for changes to the following:
```
- ensemble_calibration
- GenerateProbabilitiesFromPercentiles2D
- probability_of_snow_falling_level (special case)
- Removing the source_realizations attribute
```

### CODE REVIEW

File: lib/improver/tests/utilities/test_cube_extraction.py [hunk 1]
```diff
@@ -219,7 +222,8 @@ def test_multiple_constraints_with_units(self):
     def test_error_non_coord_units(self):
         """ Test error raised if units are provided for a non-coordinate
         constraint """
-        constraint_dict = {"name": "probability_of_precipitation"}
+        constraint_dict = {
+            "name": "probability_of_precipitation_above_threshold"}
```
Comment (bayliffe) [inline lib/improver/tests/utilities/test_cube_extraction.py]: ```probability_of_precipitation_rate_above_threshold```

File: tests/improver-wxcode/01-help.bats [hunk 2]
```diff
@@ -41,21 +41,21 @@ Calculate gridded weather symbol codes.
...
- - probability_of_cloud_area_fraction_assuming_only_consider_surface_to_1000_feet_asl; thresholds: above 0.85 (1)
- - probability_of_rainfall_rate_in_vicinity; thresholds: above 0.1 (mm hr-1), above 1.0 (mm hr-1)
- - probability_of_lwe_snowfall_rate_in_vicinity; thresholds: above 0.1 (mm hr-1), above 1.0 (mm hr-1)
+ - probability_of_rainfall_rate_above_threshold; thresholds: above 0.03 (mm hr-1), above 0.1 (mm hr-1), above 1.0 (mm hr-1)
+ - probability_of_lwe_snowfall_rate_above_threshold; thresholds: above 0.1 (mm hr-1), above 1.0 (mm hr-1)
+ - probability_of_cloud_area_fraction_above_threshold; thresholds: above 0.1875 (1), above 0.8125 (1)
+ - probability_of_visibility_in_air_below_threshold; thresholds: below 1000.0 (m), below 5000.0 (m)
+ - probability_of_cloud_area_fraction_assuming_only_consider_surface_to_1000_feet_asl_above_threshold; thresholds: above 0.85 (1)
+ - probability_of_rainfall_rate_in_vicinity_above_threshold; thresholds: above 0.1 (mm hr-1), above 1.0 (mm hr-1)
+ - probability_of_lwe_snowfall_rate_in_vicinity_above_threshold; thresholds: above 0.1 (mm hr-1), above 1.0 (mm hr-1)
```
Comment (bayliffe) [inline tests/improver-wxcode/01-help.bats]: We could now get rid of the "above" and "below" that precede each value above. The name of the diagnostic makes it clear in which sense we are interested relative to these thresholds, e.g.
```
probability_of_visibility_in_air_below_threshold; thresholds: 1000.0 (m), 5000.0 (m)
```

File: tests/improver-wxcode/01-help.bats [hunk 3]
```diff
@@ -41,21 +41,21 @@ Calculate gridded weather symbol codes.
...
- - probability_of_rainfall_rate; thresholds: above 0.03 (mm hr-1), above 0.1 (mm hr-1), above 1.0 (mm hr-1)
- - probability_of_lwe_snowfall_rate; thresholds: above 0.1 (mm hr-1), above 1.0 (mm hr-1)
- - probability_of_cloud_area_fraction; thresholds: above 0.1875 (1), above 0.8125 (1)
- - probability_of_visibility_in_air; thresholds: below 1000.0 (m), below 5000.0 (m)
- - probability_of_cloud_area_fraction_assuming_only_consider_surface_to_1000_feet_asl; thresholds: above 0.85 (1)
+ - probability_of_rainfall_rate_above_threshold; thresholds: above 0.03 (mm hr-1), above 0.1 (mm hr-1), above 1.0 (mm hr-1)
+ - probability_of_lwe_snowfall_rate_above_threshold; thresholds: above 0.1 (mm hr-1), above 1.0 (mm hr-1)
+ - probability_of_cloud_area_fraction_above_threshold; thresholds: above 0.1875 (1), above 0.8125 (1)
+ - probability_of_visibility_in_air_below_threshold; thresholds: below 1000.0 (m), below 5000.0 (m)
+ - probability_of_cloud_area_fraction_assuming_only_consider_surface_to_1000_feet_asl_above_threshold; thresholds: above 0.85 (1)
```
Comment (bayliffe) [inline tests/improver-wxcode/01-help.bats]: As above.

File: lib/improver/threshold.py [hunk 4]
```diff
@@ -272,9 +272,12 @@ def process(self, input_cube):
         # Force the metadata to temporary conventions
         if self.below_thresh_ok:
             cube.attributes.update({'relative_to_threshold': 'below'})
+            cube.rename(
+                "probability_of_{}_below_threshold".format(cube.name()))
         else:
             cube.attributes.update({'relative_to_threshold': 'above'})
-        cube.rename("probability_of_{}".format(cube.name()))
+            cube.rename(
+                "probability_of_{}_above_threshold".format(cube.name()))
```
Comment (bayliffe) [inline lib/improver/threshold.py]: The ```Returns``` section of the ```process``` doc string needs to be updated to reflect the changes that have been made here.

File: lib/improver/nowcasting/lightning.py [hunk 5]
```diff
@@ -186,7 +186,7 @@ def _update_metadata(cube):
         Modify the meta data of input cube to resemble a Nowcast of lightning
         probability.
 
-        1. Rename to "probability_of_lightning"
+        1. Rename to "probability_of_lightning_above_threshold"
```
[the remark this answers is not in the document]
Comment (markysparks) [reply lib/improver/nowcasting/lightning.py]: Fixed thanks - I think my eyes must have been glazing over after all these 'probability_of....' statements!

File: lib/improver/tests/blending/weighted_blend/test_WeightedBlendAcrossWholeDimension.py [hunk 6]
```diff
@@ -183,8 +183,9 @@ def setUp(self):
         data_threshold[1, 1, :, :] = 0.6
         data_threshold[1, 2, :, :] = 0.8
 
-        cube_threshold = Cube(data_threshold,
-                              long_name="probability_of_precipitation_amount")
+        cube_threshold = Cube(data_threshold, long_name="probability_of_"
+                                                        "precipitation_amount"
+                                                        "_below_threshold")
```
[the remark this answers is not in the document]
Comment (markysparks) [reply lib/improver/tests/blending/weighted_blend/test_WeightedBlendAcrossWholeDimension.py]: Fixed.

File: lib/improver/tests/nowcasting/lightning/test_NowcastLightning.py [hunk 7]
```diff
@@ -115,7 +115,8 @@ def test_basic(self):
         cell methods to be discarded."""
         result = self.plugin._update_metadata(self.cube)
         self.assertIsInstance(result, Cube)
-        self.assertEqual(result.name(), "probability_of_lightning")
+        self.assertEqual(
+            result.name(), "probability_of_lightning_above_threshold")
```
[the remark this answers is not in the document]
Comment (markysparks) [reply lib/improver/tests/nowcasting/lightning/test_NowcastLightning.py]: fixed.

File: lib/improver/tests/nowcasting/lightning/test_NowcastLightning.py [hunk 8]
```diff
@@ -451,7 +452,7 @@ def setUp(self):
         threshold_coord.points = [0.5, 7.0, 35.0]
         threshold_coord.rename('threshold')
         threshold_coord.units = cf_units.Unit('mm hr-1')
-        self.precip_cube.rename("probability_of_precipitation")
+        self.precip_cube.rename("probability_of_precipitation_above_threshold")
```
[the remark this answers is not in the document]
Comment (markysparks) [reply lib/improver/tests/nowcasting/lightning/test_NowcastLightning.py]: fixed

File: lib/improver/tests/utilities/test_cube_extraction.py [hunk 9]
```diff
@@ -58,15 +58,16 @@ def set_up_precip_probability_cube():
...
+    threshold = DimCoord(MMH_TO_MS * np.array([0.03, 0.1, 1.0]),
                          long_name="threshold", units="m s-1")
     ycoord = DimCoord(np.arange(3), "projection_y_coordinate")
     xcoord = DimCoord(np.arange(3), "projection_x_coordinate")
 
-    cube = iris.cube.Cube(data, long_name="probability_of_precipitation",
-                          dim_coords_and_dims=[(threshold, 0), (ycoord, 1),
-                                               (xcoord, 2)], units="1")
+    cube = iris.cube.Cube(
+        data, long_name="probability_of_precipitation_above_threshold",
```
[the remark this answers is not in the document]
Comment (markysparks) [reply lib/improver/tests/utilities/test_cube_extraction.py]: fixed

File: lib/improver/nowcasting/lightning.py [hunk 10]
```diff
@@ -465,16 +465,16 @@ def process(self, cubelist):
...
         """
         first_guess_lightning_cube = cubelist.extract(
-            "probability_of_lightning", strict=True)
+            "probability_of_lightning_above_threshold", strict=True)
         lightning_rate_cube = cubelist.extract(
             "rate_of_lightning", strict=True)
         lightning_rate_cube.convert_units("min^-1")  # Ensure units are correct
         prob_precip_cube = cubelist.extract(
-            "probability_of_precipitation", strict=True)
+            "probability_of_precipitation_above_threshold", strict=True)
```
[the remark this answers is not in the document]
Comment (markysparks) [reply lib/improver/nowcasting/lightning.py]: fixed

File: lib/improver/tests/nowcasting/lightning/test_NowcastLightning.py [hunk 11]
```diff
@@ -427,21 +428,21 @@ def setUp(self):
...
-            Cube name is "probability_of_lightning".
+            Cube name is "probability_of_lightning_above_threshold".
         self.precip_cube:
             With extra coordinate of length(3) "threshold" containing
             points [0.5, 7., 35.] mm hr-1.
             All points contain float(1.) except the
             zero point [0, 0, 1, 1] which is float(0.)
             and [1:, 0, ...] which are float(0.)
-            Cube name is "probability_of_precipitation".
+            Cube name is "probability_of_precipitation_above_threshold".
```
[the remark this answers is not in the document]
Comment (markysparks) [reply lib/improver/tests/nowcasting/lightning/test_NowcastLightning.py]: fixed

File: lib/improver/tests/nowcasting/lightning/test_NowcastLightning.py [hunk 12]
```diff
@@ -786,7 +790,7 @@ def setUp(self):
         threshold_coord.points = [0.5, 7.0, 35.0]
         threshold_coord.rename('threshold')
         threshold_coord.units = cf_units.Unit('mm hr-1')
-        self.precip_cube.rename("probability_of_precipitation")
+        self.precip_cube.rename("probability_of_precipitation_above_threshold")
```
[the remark this answers is not in the document]
Comment (markysparks) [reply lib/improver/tests/nowcasting/lightning/test_NowcastLightning.py]: fixed

File: lib/improver/tests/utilities/test_cube_extraction.py [hunk 13]
```diff
@@ -189,7 +190,8 @@ def setUp(self):
 
     def test_basic_no_units(self):
         """ Test cube extraction for single constraint without units """
-        constraint_dict = {"name": "probability_of_precipitation"}
+        constraint_dict = {
+            "name": "probability_of_precipitation_above_threshold"}
```
[the remark this answers is not in the document]
Comment (markysparks) [reply lib/improver/tests/utilities/test_cube_extraction.py]: fixed

File: lib/improver/tests/utilities/test_cube_extraction.py [hunk 14]
```diff
@@ -208,8 +210,9 @@ def test_basic_with_units(self):
 
     def test_multiple_constraints_with_units(self):
         """ Test behaviour with a list of constraints and units """
-        constraint_dict = {"name": "probability_of_precipitation",
-                           "threshold": 0.03}
+        constraint_dict = {
+            "name": "probability_of_precipitation_above_threshold",
```
[the remark this answers is not in the document]
Comment (markysparks) [reply lib/improver/tests/utilities/test_cube_extraction.py]: fixed

File: lib/improver/tests/utilities/test_cube_extraction.py [hunk 15]
```diff
@@ -228,8 +232,9 @@ def test_return_none(self):
         """ Test function returns None rather than raising an error where
         no subcubes match the required constraints, when unit conversion is
         required """
-        constraint_dict = {"name": "probability_of_precipitation",
-                           "threshold": 5}
+        constraint_dict = {
+            "name": "probability_of_precipitation_above_threshold",
```
[the remark this answers is not in the document]
Comment (markysparks) [reply lib/improver/tests/utilities/test_cube_extraction.py]: fixed

File: lib/improver/tests/utilities/test_cube_extraction.py (hunk 1 above)
Comment (markysparks) [reply lib/improver/tests/utilities/test_cube_extraction.py]: fixed

File: lib/improver/threshold.py (hunk 4 above)
Comment (markysparks) [reply lib/improver/threshold.py]: Updated

File: tests/improver-wxcode/01-help.bats (hunk 2 above)
Comment (markysparks) [reply tests/improver-wxcode/01-help.bats]: Good point, updated

File: tests/improver-wxcode/01-help.bats (hunk 3 above)
Comment (markysparks) [reply tests/improver-wxcode/01-help.bats]: Updated

File: lib/improver/tests/set_up_test_cubes.py [hunk 16]
```diff
@@ -366,9 +366,15 @@ def set_up_probability_cube(data, thresholds, variable_name='air_temperature',
...
-        name = variable_name
+        name_prefix = ''
     else:
-        name = 'probability_of_{}'.format(variable_name)
+        name_prefix = 'probability_of_'
+
+    if relative_to_threshold == 'above':
+        name = name_prefix + '{}_above_threshold'.format(variable_name)
+    else:
+        name = name_prefix + '{}_below_threshold'.format(variable_name)
```
[the remark this answers is not in the document]
Comment (markysparks) [reply lib/improver/tests/set_up_test_cubes.py]: Changed as requested. I have written a new test to check the exception is raised.

**Comment by gavinevans** [comment]
Just to highlight, that there will likely be conflicts between this PR and #814.

File: lib/improver/tests/test_set_up_test_cubes.py [hunk 17]
```diff
@@ -361,6 +361,15 @@ def test_relative_to_threshold(self):
         self.assertEqual(len(result.attributes), 1)
         self.assertEqual(result.attributes['relative_to_threshold'], 'below')
 
+    def test_relative_to_threshold_set(self):
+        """Test that an error is raised if the "relative_to_threshold"
+        attribute has not been set when setting up a probability cube"""
+        data = np.flipud(self.data)
```
Comment (bayliffe) [inline lib/improver/tests/test_set_up_test_cubes.py]: We don't need this np.flipud as we are not trying to invert the relative_to_threshold attribute, just set it to none.

**Comment by bayliffe** [comment]
Only one small comment. Otherwise I am happy with the changes. Good work Mark.

File: lib/improver/tests/test_set_up_test_cubes.py (hunk 17 above)
Comment (markysparks) [reply lib/improver/tests/test_set_up_test_cubes.py]: oops - removed

File: lib/improver/nowcasting/lightning.py [hunk 18]
```diff
@@ -186,7 +186,7 @@ def _update_metadata(cube):
         Modify the meta data of input cube to resemble a Nowcast of lightning
         probability.
 
-        1. Rename to "probability_of_lightning"
+        1. Rename to "probability_of_lightning_rate_above_threshold"
```
Comment (fionaRust) [inline lib/improver/nowcasting/lightning.py]: @MoseleyS  Will this cause any problems with the old operational system (am I right the code is used in gpp?)

File: lib/improver/tests/set_up_test_cubes.py [hunk 19]
```diff
@@ -366,9 +366,19 @@ def set_up_probability_cube(data, thresholds, variable_name='air_temperature',
         attributes['relative_to_threshold'] = relative_to_threshold
 
     if variable_name.startswith('probability_of_'):
-        name = variable_name
+        name_prefix = ''
     else:
-        name = 'probability_of_{}'.format(variable_name)
+        name_prefix = 'probability_of_'
```
Comment (fionaRust) [inline lib/improver/tests/set_up_test_cubes.py]: You could also update the Docstring where it mentions the probability_of_X convention

File: lib/improver/threshold.py [hunk 20]
```diff
@@ -198,7 +198,9 @@ def process(self, input_cube):
                 a given threshold has been exceeded or not.
 
                 The cube meta-data will contain:
-                 * input_cube name prepended with `probability_of_`
+                 * input_cube name prepended with
```
[the remark this answers is not in the document]
Comment (markysparks) [reply lib/improver/threshold.py]: Fixed.

File: lib/improver/tests/set_up_test_cubes.py (hunk 19 above)
Comment (markysparks) [reply lib/improver/tests/set_up_test_cubes.py]: Updated.

**Comment by fionaRust** [comment]
This is ready but waiting for all PRs then merging all together.

**Pull request merged**